### PR TITLE
[UI-side compositing] REGRESSION: empty Mail compose window always shows a scrollbar

### PR DESCRIPTION
**Pull request by nmoucht**
#### e41f5679593f6e0a4eb380dd580392234ea3fd4b
<pre>
[UI-side compositing] REGRESSION: empty Mail compose window always shows a scrollbar
<a href="https://bugs.webkit.org/show_bug.cgi?id=259323">https://bugs.webkit.org/show_bug.cgi?id=259323</a>
rdar://110872395

Reviewed by Simon Fraser.

To properly set NSScrollerImps as enabled in the UI-process, send accross the web
process state from the Scrollbar class. It is also necessary to set the state in
setScrollingNodeScrollableAreaGeometry because the first time setting this state
on the Scrollbar class can happen before the corresponding scrollable area has been
added to the scrolling tree.

* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
(WebCore::AsyncScrollingCoordinator::setScrollbarEnabled):
(WebCore::AsyncScrollingCoordinator::setScrollingNodeScrollableAreaGeometry):
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.h:
* Source/WebCore/page/scrolling/ScrollingCoordinator.h:
(WebCore::ScrollingCoordinator::setScrollbarEnabled):
* Source/WebCore/page/scrolling/ScrollingStateNode.h:
* Source/WebCore/page/scrolling/ScrollingStateScrollingNode.cpp:
(WebCore::ScrollingStateScrollingNode::ScrollingStateScrollingNode):
(WebCore::ScrollingStateScrollingNode::setScrollbarEnabledState):
* Source/WebCore/page/scrolling/ScrollingStateScrollingNode.h:
(WebCore::ScrollingStateScrollingNode::scrollbarEnabledState const):
* Source/WebCore/page/scrolling/mac/ScrollerMac.h:
(WebCore::ScrollerMac::setEnabled):
* Source/WebCore/page/scrolling/mac/ScrollerMac.mm:
(WebCore::ScrollerMac::updateValues):
* Source/WebCore/page/scrolling/mac/ScrollingTreeScrollingNodeDelegateMac.mm:
(WebCore::ScrollingTreeScrollingNodeDelegateMac::updateFromStateNode):
* Source/WebCore/platform/Scrollbar.cpp:
(WebCore::Scrollbar::setEnabled):
* Source/WebCore/platform/ScrollbarsController.h:
(WebCore::ScrollbarsController::setScrollbarEnabledSate):
* Source/WebKit/Shared/RemoteLayerTree/RemoteScrollingCoordinatorTransaction.cpp:
(ArgumentCoder&lt;ScrollingStateScrollingNode&gt;::encode):
(ArgumentCoder&lt;ScrollingStateScrollingNode&gt;::decode):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollbarsController.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollbarsController.mm:
(WebKit::RemoteScrollbarsController::setScrollbarEnabledSate):

Canonical link: <a href="https://commits.webkit.org/266261@main">https://commits.webkit.org/266261@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/6eebf02527793c85606fb5508368b5fdd2c5bf63

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/13254 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/13566 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/13901 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/14991 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/12617 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/13333 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/16075 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/13594 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/15306 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/13421 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/14082 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/11202 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/15593 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/11370 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/11957 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/19014 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/12445 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/12124 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/15337 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/12622 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/10482 "1 failures") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/11893 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/16215 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1525 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/12464 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp [hunk 1]
```diff
@@ -417,6 +417,20 @@ void AsyncScrollingCoordinator::setMouseMovedInContentArea(ScrollableArea& scrol
     stateNode->setMouseMovedInContentArea(state);
 }
 
+void AsyncScrollingCoordinator::setScrollbarEnabled(Scrollbar* scrollbar)
```
Comment (smfr) [inline Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp]: Use a `Scrollbar&` parameter?

File: Source/WebCore/page/scrolling/ScrollingStateScrollingNode.cpp [hunk 2]
```diff
@@ -302,6 +303,19 @@ void ScrollingStateScrollingNode::setScrollbarHoverState(ScrollbarHoverState hov
     setPropertyChanged(Property::ScrollbarHoverState);
 }
 
+void ScrollingStateScrollingNode::setScrollbarEnabledState(ScrollbarOrientation orientation, bool flag)
```
Comment (smfr) [inline Source/WebCore/page/scrolling/ScrollingStateScrollingNode.cpp]: `flag` -> `enabled`

File: Source/WebCore/platform/ScrollbarsController.h [hunk 3]
```diff
@@ -95,6 +95,7 @@ class ScrollbarsController {
 
     WEBCORE_EXPORT virtual bool shouldDrawIntoScrollbarLayer(Scrollbar&) const { return true; }
     WEBCORE_EXPORT virtual bool shouldRegisterScrollbars() const { return true; }
+    WEBCORE_EXPORT virtual void setScrollbarEnabledSate(WebCore::Scrollbar&) { }
```
Comment (smfr) [inline Source/WebCore/platform/ScrollbarsController.h]: No need for `WebCore::`

Comment (smfr) [reply Source/WebCore/platform/ScrollbarsController.h]: It's odd to have a function called `setScrollbarEnabledSate` which doesn't take the state as an argument. Should this be called `updateScrollbarEnabledState()`?

**Review by smfr — APPROVED** [review]

File: Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp (hunk 1 above)
Comment (nmoucht) [reply Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp]: Fixed.

File: Source/WebCore/page/scrolling/ScrollingStateScrollingNode.cpp (hunk 2 above)
Comment (nmoucht) [reply Source/WebCore/page/scrolling/ScrollingStateScrollingNode.cpp]: Fixed.

File: Source/WebCore/platform/ScrollbarsController.h (hunk 3 above)
Comment (nmoucht) [reply Source/WebCore/platform/ScrollbarsController.h]: Fixed.

**Pull request merged**